README.md
easy-livereload
===============

[![Build Status](https://travis-ci.com/dai-shi/easy-livereload.svg?branch=master)](https://travis-ci.com/github/dai-shi/easy-livereload)
[![npm version](https://badge.fury.io/js/easy-livereload.svg)](https://badge.fury.io/js/easy-livereload)

This is yet another library to use [livereload](http://livereload.com/)
very easily for express/node.js-based development.
It is express middleware which provides both
a livereload server and a javascript client.
It is designed to be used with
[node-dev](https://www.npmjs.com/package/node-dev)
so that restarting a server process is also possible.

The major features of this library include:

- using livereload (PROTOCOL 7) server code from the original author,
- using livereload.js client code from the original author,
- using fs.watch() to check file changes instantly,
- allowing to automatically restart server code (if invoked with node-dev), and
- all-in-one package to enable with at least one-line code.

Install
-------

```bash
npm install easy-livereload node-dev
```

Usage
-----

Minimal configuration:

```js
app.use(require('easy-livereload')());
```

Typical configuration:

```js
var path = require('path');
var express = require('express');
var app = express();

if (app.get('env') === 'development') {
  var livereload = require('easy-livereload');
  var file_type_map = {
    jade: 'html', // `index.jade` maps to `index.html`
    styl: 'css', // `styles/site.styl` maps to `styles/site.css`
    scss: 'css', // `styles/site.scss` maps to `styles/site.css`
    sass: 'css', // `styles/site.scss` maps to `styles/site.css`
    less: 'css' // `styles/site.scss` maps to `styles/site.css`
    // add the file type being edited and what you want it to be mapped to.
  };
  
  // store the generated regex of the object keys
  var file_type_regex = new RegExp('\\.(' + Object.keys(file_type_map).join('|') + ')$');
  
  app.use(livereload({
    watchDirs: [
      path.join(__dirname, 'public'),
      path.join(__dirname, 'app')
    ],
    checkFunc: function(file) {
      return file_type_regex.test(file);
    },
    renameFunc: function(file) {
      // remap extention of the file path to one of the extentions in `file_type_map`
      return file.replace(file_type_regex, function(extention) {
        return '.' + file_type_map[extention.slice(1)];
      });
    },
    port: process.env.LIVERELOAD_PORT || 35729
  }));
}
```

By default this script tries to load the live reload script itself,
but if that doesn't work for some reason then you can put your `app`
into the `easy-livereload` options.
This will add a local variable to your `app` under `app.locals.LRScript`.

```js
var express = require('express');
var app = express();
var livereload = require('easy-livereload');

if (app.get('env') === 'development') {
  app.use(livereload({
    app: app
  }));
}
```

```jade
doctype html
html(lang="en")
  head
    title Livereload
    
    != LRScript //- loads the livereload script
```


Example scripts entry in `package.json`:

```json
"scripts": {
  "start": "NODE_ENV=production node app.js",
  "start-dev": "NODE_ENV=development node-dev app.js"
}
```


CHANGELOG.md
# Change Log

## [Unreleased]

## [1.4.4] - 2022-08-31
### Changed
- Fix ECONNRESET #19

## [1.4.3] - 2020-08-27
### Changed
- Ignore ECONNRESET (#14)

## [1.4.2] - 2020-05-05
### Changed
- Fix a misusage in v1.4.1 (#11)

## [1.4.1] - 2020-05-05
### Changed
- Update node-watch version (#9)

## [1.4.0] - 2020-03-22
### Changed
- Make reloadTimeout optional (#7)

## [1.3.2] - 2018-07-23
### Changed
- Workaround for websocket error handling (#6)

## [1.3.1] - 2018-03-02
### Changed
- Fix for no content-type header (#5)

## [1.3.0] - 2017-08-31
### Changed
- Workaround for process.EventEmitter (#4)

## [1.2.0] - 2016-09-11
### Added
- npm3 support

## [1.1.0] - 2016-06-28
### Added
- Update livereload-js, obsoleting the use of napa

## [1.0.4] - 2016-01-14
### Added
- Option to reference a local variable for the LiveReload script

## [1.0.3] - 2015-01-07
### Changed
- Use of napa v1.2.0

## [1.0.2] - 2015-01-01
### Fixed
- Minor fixes

## [1.0.1] - 2015-01-01
### Added
- Lots of improvements

## [1.0.0] - 2014-12-31
### Added
- Initial release


test/10_basic_test.js
/* eslint-env mocha */

var express = require('express');
var request = require('supertest');
var livereload = require('../index.js');

var app = express();

var port = process.env.LIVERELOAD_TEST_PORT || 45729;
app.use(livereload({
  port: port,
  watchDirs: ['./']
}));

app.get('/', function(req, res) {
  res.send('<html><head><title>test</title></head><body><h1>test</h1></body></html');
});

app.listen(process.env.PORT || 27892);

describe('basic test', function() {
  it('should include livereload script code', function(done) {
    request(app)
      .get('/')
      .expect(200)
      .expect('Content-Type', /html/)
      .expect(/livereload/)
      .end(function(err) {
        if (err) return done(err);
        done();
      });
  });

  it('should have livereload server available', function(done) {
    request('http://localhost:' + port)
      .get('/')
      .expect(404)
      .end(function(err) {
        if (err) return done(err);
        done();
      });
  });

  it('should be able to load livereload.js', function(done) {
    request('http://localhost:' + port)
      .get('/livereload.js')
      .expect(200)
      .expect('Content-Type', /javascript/)
      .expect(/PROTOCOL_7/)
      .end(function(err) {
        if (err) return done(err);
        done();
      });
  });
});


index.js
/*
  Copyright (C) 2014, Daishi Kato <daishi@axlight.com>
  All rights reserved.

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
  HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

// This is a workaround for https://github.com/dai-shi/easy-livereload/issues/4.
// Instead, we should update dependent packages once they are ready.
process.EventEmitter = process.EventEmitter || require('events');

var fs = require('fs');
var path = require('path');
var LRWebSocketServer = require('livereload-server');
var watch = require('node-watch');
var bodyRewrite = require('connect-body-rewrite');

var lrserver = null;

var orgCreateConnection = LRWebSocketServer.prototype._createConnection;
LRWebSocketServer.prototype._createConnection = function(socket) {
  orgCreateConnection.call(this, socket);
  socket.on('error', function(err) {
    if (err.code === 'ECONNRESET') {
      // ignore ECONNRESET connection closed by client error
      return;
    }
    console.error(err);
  });
};

function startLRServer(options) {
  if (lrserver) return;

  options.livereload = options.livereload || {};
  lrserver = new LRWebSocketServer({
    id: options.livereload.id || 'default id',
    name: options.livereload.name || 'default name',
    version: options.livereload.version || '1.0',
    protocols: {
      monitoring: 7,
      saving: 1
    },
    port: options.port
  });

  lrserver.on('livereload.js', function(req, res) {
    fs.readFile(path.join(require.resolve('livereload-js'), '../../', 'dist', 'livereload.js'), 'utf8', function(err, data) {
      if (err) throw err;
      res.writeHead(200, {
        'Content-Length': data.length,
        'Content-Type': 'text/javascript'
      });
      res.end(data);
    });
  });

  lrserver.on('httprequest', function(url, req, res) {
    res.writeHead(404);
    res.end();
  });

  lrserver.listen(function(err) {
    if (err) throw err;
  });

  var sendAll = function(command) {
    Object.keys(lrserver.connections).forEach(function(id) {
      try {
        lrserver.connections[id].send(command);
      } catch (e) {
        console.error('Livereload send command failed: %s', id);
      }
    });
  };

  options.watchDirs.forEach(function(dir) {
    watch(dir, function(event, file) {
      file = path.relative(dir, file);
      if (options.checkFunc(file)) {
        sendAll({
          command: 'reload',
          path: options.renameFunc(file),
          liveCSS: true
        });
      }
    });
  });
}

module.exports = function(options) {
  options = options || {};
  options.host = options.host || 'localhost';
  options.port = options.port || 35729;
  options.restartTimeout = options.restartTimeout || 1000;
  options.watchDirs = options.watchDirs || ['public'];
  options.checkFunc = options.checkFunc || function(x) {
    return /\.(css|js)$/.test(x);
  };
  options.renameFunc = options.renameFunc || function(x) {
    return x;
  };

  var code = '<script>document.write(\'<script src="//\' + (location.host || \'' + options.host + '\').split(\':\')[0] + \':' + options.port + '/livereload.js?snipver=1"></\' + \'script>\')</script>';
  if (options.reloadTimeout > 0) {
    code += '<script>document.addEventListener(\'LiveReloadDisconnect\', function() { setTimeout(function() { window.location.reload(); }, ' + options.reloadTimeout + '); })</script>';
  }

  if (options.app) {
    options.app.locals.LRScript = code;
  }

  if (options.restartTimeout > 0) {
    process.on('SIGTERM', function() {
      setTimeout(function() {
        process.exit(0);
      }, options.restartTimeout);
    });
  }

  startLRServer(options);

  return bodyRewrite({
    accept: function(res) {
      return /text\/html/.test(res.getHeader('content-type'));
    },
    rewrite: function(body) {
      return body.replace(/<\/body>/, code + '</body>');
    }
  });
};
